File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/API/TosoGameAPI.java
package jp.aoichaan0513.A_TosoGame_Live.API;

import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MissionManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager;
import jp.aoichaan0513.A_TosoGame_Live.API.Maps.MapUtility;
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import jp.aoichaan0513.A_TosoGame_Live.Utils.DateTime.TimeFormat;
import org.bukkit.*;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class TosoGameAPI {

    public static boolean isRes = true;
    // public static HashMap<UUID, Integer> respawnCountMap = new HashMap<>();

    // 復活クールタイム
    // public static HashMap<UUID, Integer> respawnCoolTimeMap = new HashMap<>();
    public static HashMap<UUID, WorldManager.Difficulty> difficultyMap = new HashMap<>();

    public static void setItem(WorldManager.GameType type, Player p) {
        if (GameManager.isGame(GameManager.GameState.GAME)) {
            if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p)) {
                if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
                    Inventory i = p.getInventory();
                    i.clear();

                    ItemStack infoStack = new ItemStack(Material.GOLD_NUGGET);
                    ItemMeta infoMeta = infoStack.getItemMeta();
                    infoMeta.setDisplayName(ChatColor.GREEN + "プレイヤー情報");
                    infoMeta.setLore(Arrays.asList(ChatColor.YELLOW + "プレイヤーを右クリックして", ChatColor.YELLOW + "そのプレイヤーの情報が確認できます。"));
                    infoStack.setItemMeta(infoMeta);

                    i.setItem(1, infoStack);

                    if (Bukkit.getPluginManager().isPluginEnabled("WorldEdit")) {
                        ItemStack editStack = new ItemStack(Material.COMPASS);
                        ItemMeta editMeta = editStack.getItemMeta();
                        editMeta.setDisplayName(ChatColor.GREEN + "コンパス (WorldEdit)");
                        editMeta.setLore(Arrays.asList(ChatColor.YELLOW + "右クリックした方向にテレポートができます。"));
                        editStack.setItemMeta(editMeta);

                        i.setItem(2, editStack);
                    }

                    MissionManager.reloadBook(p);
                    return;
                } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_JAIL, p)) {
                    Inventory i = p.getInventory();
                    i.clear();

                    MissionManager.reloadBook(p);

                    ItemStack itemStack = new ItemStack(Material.ENDER_PEARL);
                    ItemMeta itemMeta = itemStack.getItemMeta();
                    itemMeta.setDisplayName(ChatColor.GREEN + "プレイヤーを非表示");
                    itemMeta.setLore(Arrays.asList(ChatColor.YELLOW + "右クリックして運営以外のプレイヤーを非表示にします。"));
                    itemStack.setItemMeta(itemMeta);

                    i.setItem(8, itemStack);
                    return;
                } else {
                    WorldConfig worldConfig = Main.getWorldConfig();
                    WorldConfig.DifficultyConfig difficultyConfig = worldConfig.getDifficultyConfig(p);

                    Inventory i = p.getInventory();
                    i.clear();

                    if (MapUtility.getMap() != null) {
                        ItemStack mapStack = MapUtility.getMap().clone();
                        if (mapStack != null) {
                            ItemMeta mapMeta = mapStack.getItemMeta();
                            mapMeta.setDisplayName(ChatColor.GREEN + "地図");
                            mapMeta.setLore(Arrays.asList(ChatColor.YELLOW + "なんかすごいやつ (語彙力)"));
                            mapStack.setItemMeta(mapMeta);

                            i.setItem(40, mapStack);
                        }
                    }

                    WorldConfig.DifficultyConfig.IItem boneItem = difficultyConfig.getBone(type);
                    ItemStack boneStack = new ItemStack(Material.BONE, boneItem.getCount());
                    ItemMeta boneMeta = boneStack.getItemMeta();
                    boneMeta.setDisplayName(ChatColor.GREEN + "骨 (透明化)");
                    boneMeta.setLore(
                            Arrays.asList(
                                    ChatColor.YELLOW + "クリックで" + TimeFormat.formatSec(boneItem.getDuration()) + "秒間透明になります。",
                                    "",
                                    "" + ChatColor.BLUE + ChatColor.UNDERLINE + "クールタイム" + ChatColor.RESET + ChatColor.GRAY + ": " + ChatColor.YELLOW + TimeFormat.formatSec(boneItem.getDuration() + 5) + ChatColor.GRAY + "秒"
                            )
                    );
                    boneStack.setItemMeta(boneMeta);

                    WorldConfig.DifficultyConfig.IItem featherItem = difficultyConfig.getFeather(type);
                    ItemStack featherStack = new ItemStack(Material.FEATHER, featherItem.getCount());
                    ItemMeta featherMeta = featherStack.getItemMeta();
                    featherMeta.setDisplayName(ChatColor.GREEN + "羽 (移動速度上昇)");
                    featherMeta.setLore(
                            Arrays.asList(
                                    ChatColor.YELLOW + "クリックで" + TimeFormat.formatSec(featherItem.getDuration()) + "秒間移動速度が上昇します。",
                                    "",
                                    "" + ChatColor.BLUE + ChatColor.UNDERLINE + "クールタイム" + ChatColor.RESET + ChatColor.GRAY + ": " + ChatColor.YELLOW + TimeFormat.formatSec(featherItem.getDuration() + 5) + ChatColor.GRAY + "秒"
                            )
                    );
                    featherStack.setItemMeta(featherMeta);

                    ItemStack eggStack = new ItemStack(Material.EGG, difficultyConfig.getEgg(type).getCount());
                    ItemMeta eggMeta = eggStack.getItemMeta();
                    eggMeta.setDisplayName(ChatColor.GREEN + "卵 (盲目・移動速度低下)");
                    eggMeta.setLore(Arrays.asList(ChatColor.YELLOW + "ハンターに当てると盲目と移動速度低下を5秒間与えます。"));
                    eggStack.setItemMeta(eggMeta);

                    i.setItem(0, boneStack);
                    i.setItem(1, featherStack);
                    i.setItem(2, eggStack);

                    MissionManager.reloadBook(p);
                    return;
                }
            }
        } else {
            if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
                Inventory i = p.getInventory();
                i.clear();

                ItemStack infoStack = new ItemStack(Material.GOLD_NUGGET);
                ItemMeta infoMeta = infoStack.getItemMeta();
                infoMeta.setDisplayName(ChatColor.GREEN + "プレイヤー情報");
                infoMeta.setLore(Arrays.asList(ChatColor.YELLOW + "プレイヤーを右クリックして", ChatColor.YELLOW + "そのプレイヤーの情報が確認できます。"));
                infoStack.setItemMeta(infoMeta);

                i.setItem(1, infoStack);

                if (Bukkit.getPluginManager().isPluginEnabled("WorldEdit")) {
                    ItemStack editStack = new ItemStack(Material.COMPASS);
                    ItemMeta editMeta = editStack.getItemMeta();
                    editMeta.setDisplayName(ChatColor.GREEN + "コンパス (WorldEdit)");
                    editMeta.setLore(Arrays.asList(ChatColor.YELLOW + "クリックした方向にテレポートができます。"));
                    editStack.setItemMeta(editMeta);

                    i.setItem(2, editStack);
                }

                MissionManager.reloadBook(p);
                return;
            } else {
                p.getInventory().clear();
                MissionManager.reloadBook(p);
                return;
            }
        }
    }

    public static void setArmor(Player p) {
        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p)) {
            p.getInventory().clear();
            p.getInventory().setHelmet(new ItemStack(Material.DIAMOND_HELMET));
            p.getInventory().setChestplate(new ItemStack(Material.DIAMOND_CHESTPLATE));
            p.getInventory().setLeggings(new ItemStack(Material.DIAMOND_LEGGINGS));
            p.getInventory().setBoots(new ItemStack(Material.DIAMOND_BOOTS));
        } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_TUHO, p)) {
            p.getInventory().clear();
            // p.getInventory().setHelmet(new ItemStack(Material.GOLDEN_HELMET));
            p.getInventory().setHelmet(new ItemStack(Material.HONEY_BLOCK));
            p.getInventory().setChestplate(new ItemStack(Material.GOLDEN_CHESTPLATE));
            p.getInventory().setLeggings(new ItemStack(Material.GOLDEN_LEGGINGS));
            p.getInventory().setBoots(new ItemStack(Material.GOLDEN_BOOTS));
        }
    }

    public static void removeArmor(Player p) {
        p.getInventory().setHelmet(new ItemStack(Material.AIR));
        p.getInventory().setChestplate(new ItemStack(Material.AIR));
        p.getInventory().setLeggings(new ItemStack(Material.AIR));
        p.getInventory().setBoots(new ItemStack(Material.AIR));
    }

    public static void setPotionEffect(Player p) {
        setPotionEffect(p, true);
    }

    public static void setPotionEffect(Player p, boolean isRemove) {
        if (isRemove)
            for (PotionEffect effect : p.getActivePotionEffects())
                p.removePotionEffect(effect.getType());

        p.addPotionEffect(new PotionEffect(PotionEffectType.NIGHT_VISION, 200000, 1, false, false));

        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p)) {
            if (GameManager.isGame(GameManager.GameState.READY) && GameManager.isGame(GameManager.GameState.GAME)) {
                p.addPotionEffect(new PotionEffect(PotionEffectType.INVISIBILITY, 200, 1, false, false));
                p.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 200, 1, false, false));
            }
        } else if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_HUNTER, p)) {
            p.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 200000, 0, false, false));
        }
    }

    public static void showPlayers(Player p) {
        for (Player player : Bukkit.getOnlinePlayers()) {
            if (MainAPI.isHidePlayer(player)) {
                // disappear コマンドで非表示にしていた時
                p.hidePlayer(player);
            } else {
                // 配信者が非表示設定にしていた時
                if (isHidePlayer(player)) {
                    if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p))
                        p.hidePlayer(player);
                    else
                        p.showPlayer(player);
                } else {
                    p.showPlayer(player);
                }
            }
        }

        // 逃走者が透明を使用した時
        if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p)) {
            for (Player player : MainAPI.getOnlinePlayers(Main.invisibleSet))
                p.hidePlayer(player);
        } else {
            for (Player player : MainAPI.getOnlinePlayers(Main.invisibleSet))
                p.showPlayer(player);
        }
    }

    public static boolean isPlayer(CommandSender sender) {
        return sender instanceof Player;
    }

    public static boolean isAdmin(Player p) {
        return p.isOp() && (isBroadCaster(p) || hasPermission(p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p));
    }

    public static boolean hasPermission(Player p) {
        return PlayerManager.loadConfig(p).getPermission();
    }

    public static boolean isBroadCaster(Player p) {
        return PlayerManager.loadConfig(p).getBroadCaster();
    }

    public static void toggleOp(Player p) {
        if (!isBroadCaster(p) && !hasPermission(p)) return;
        if (p.isOp())
            removeOp(p);
        else
            addOp(p);
    }

    public static void addOp(Player p) {
        if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p) || hasPermission(p) || isBroadCaster(p))
            p.setOp(true);
    }

    public static void removeOp(Player p) {
        if (!Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_ADMIN, p) || hasPermission(p) || isBroadCaster(p))
            p.setOp(false);
    }

    public static void teleport(Player p, Location location) {
        Bukkit.getScheduler().runTask(Main.getInstance(), () -> p.teleport(location));
    }

    public static void teleport(Player p, Collection<Location> collection) {
        Bukkit.getScheduler().runTask(Main.getInstance(), () -> {
            List<Location> list = new ArrayList<>(collection);
            Collections.shuffle(list, ThreadLocalRandom.current());
            p.teleport(list.get(0));
        });
    }

    public static void sendNotificationSound() {
        for (Player p : Bukkit.getOnlinePlayers())
            TosoGameAPI.sendNotificationSound(p);
    }

    public static void sendNotificationSound(Player p) {
        /*
        new BukkitRunnable() {
            int count = 0;

            public void run() {
                count++;
                if (count == 10 || count == 11 || count == 12 || count == 13 || count == 14 || count == 15) {
                } else if (count == 25) {
                    cancel();
                } else {
                    p.playSound(p.getLocation(), Sound.BLOCK_NOTE_BLOCK_PLING, 2, 2);
                }
            }
        }.runTaskTimer(Main.getInstance(), 3, 3);
        */
        new BukkitRunnable() {
            int c = 26;

            @Override
            public void run() {
                if ((c <= 26 && c >= 18) || (c <= 9 && c > 0))
                    p.playSound(p.getLocation(), Sound.BLOCK_NOTE_BLOCK_PLING, 1, 2);
                if (c < 0)
                    cancel();
                c--;
            }
        }.runTaskTimerAsynchronously(Main.getInstance(), 0, 2);
    }

    private static List<UUID> liveHidePlayerList = new ArrayList<>();

    public static void hidePlayers(Player p) {
        for (UUID uuid : liveHidePlayerList) {
            Player player = Bukkit.getPlayer(uuid);

            if (player == null) continue;
            if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, p) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, p))
                p.hidePlayer(player);
            else
                p.showPlayer(player);
        }
    }

    public static List<UUID> getHidePlayerList() {
        return liveHidePlayerList;
    }

    public static boolean isHidePlayer(Player p) {
        return liveHidePlayerList.contains(p.getUniqueId());
    }

    public static void addHidePlayer(Player p) {
        if (liveHidePlayerList.contains(p.getUniqueId())) return;
        liveHidePlayerList.add(p.getUniqueId());

        for (Player player : Bukkit.getOnlinePlayers()) {
            if (Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_PLAYER, player) || Teams.hasJoinedTeam(Teams.OnlineTeam.TOSO_SUCCESS, player))
                hidePlayers(player);
        }
    }

    public static void removeHidePlayer(Player p) {
        if (!liveHidePlayerList.contains(p.getUniqueId())) return;
        liveHidePlayerList.remove(p.getUniqueId());

        for (Player player : Bukkit.getOnlinePlayers()) {
            player.showPlayer(p);
            hidePlayers(player);
            MainAPI.hidePlayers(player);
        }
    }


    public enum Objective {
        // サイドバー
        SIDEBAR("Toso_Sidebar"),

        // サイドバー 表示用チーム
        SIDEBAR_STATUS("T_SB_Status"),
        SIDEBAR_TEAM("T_SB_Team"),
        SIDEBAR_DIFFICULTY("T_SB_Difficulty"),
        SIDEBAR_REWARD("T_SB_Reward"),
        SIDEBAR_RATE("T_SB_Rate"),

        SIDEBAR_TEAM_PLAYER("T_SB_T_Player"),
        SIDEBAR_TEAM_HUNTER("T_SB_T_Hunter"),
        SIDEBAR_TEAM_JAIL("T_SB_T_Jail"),
        SIDEBAR_TEAM_SUCCESS("TSB_T_Success"),
        SIDEBAR_TEAM_TUHO("T_SB_T_Tuho"),

        // チーム
        TEAM_ADMIN("Toso_Admin"),
        TEAM_PLAYER("Toso_Player"),
        TEAM_HUNTER("Toso_Hunter"),
        TEAM_JAIL("Toso_Jail"),
        TEAM_SUCCESS("Toso_Success"),
        TEAM_TUHO("Toso_Tuho");

        private final String name;

        private Objective(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/API/TosoGameAPI.kt
package jp.aoichaan0513.A_TosoGame_Live.API

import jp.aoichaan0513.A_TosoGame_Live.API.Enums.ItemType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.GameManager.GameState
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.MoneyManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.DifficultyManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerConfig
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.PlayerManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player.VisibilityManager
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager.GameType
import jp.aoichaan0513.A_TosoGame_Live.API.Map.MapUtility
import jp.aoichaan0513.A_TosoGame_Live.API.Scoreboard.Teams
import jp.aoichaan0513.A_TosoGame_Live.Main
import jp.aoichaan0513.A_TosoGame_Live.Mission.MissionManager
import jp.aoichaan0513.A_TosoGame_Live.Utils.*
import org.bukkit.*
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.scheduler.BukkitRunnable
import java.util.concurrent.ThreadLocalRandom

class TosoGameAPI {
    companion object {

        var isRespawn = true
        var isRunnedBonusMission = false

        // アイテム配布
        fun setItem(type: GameType, p: Player) {
            val inv = p.inventory

            inv.clear()

            if (p.isAdminTeam) {
                val infoStack = ItemStack(Material.GOLD_NUGGET)
                val infoMeta = infoStack.itemMeta!!
                infoMeta.setDisplayName(ItemUtil.getItemName("プレイヤー情報"))
                infoMeta.lore = listOf("${ChatColor.YELLOW}プレイヤーを右クリックして", "${ChatColor.YELLOW}そのプレイヤーの情報が確認できます。")
                infoStack.itemMeta = infoMeta
                inv.setItem(1, infoStack)

                if (Bukkit.getPluginManager().isPluginEnabled("WorldEdit")) {
                    val editStack = ItemStack(Material.COMPASS)
                    val editMeta = editStack.itemMeta!!
                    editMeta.setDisplayName(ItemUtil.getItemName("コンパス (WorldEdit)"))
                    editMeta.lore = listOf("${ChatColor.YELLOW}右クリックした方向にテレポートができます。")
                    editStack.itemMeta = editMeta
                    inv.setItem(2, editStack)
                }

                MissionManager.setBook(p)
            } else {
                if (GameManager.isGame(GameState.GAME)) {
                    when {
                        p.isHunterGroup -> {
                            setArmor(p)
                        }
                        p.isJailTeam -> {
                            MissionManager.setBook(p)
                            val itemStack = ItemStack(Material.ENDER_PEARL)
                            val itemMeta = itemStack.itemMeta!!
                            itemMeta.setDisplayName(ItemUtil.getItemName("プレイヤーを非表示"))
                            itemMeta.lore = listOf("${ChatColor.YELLOW}クリックして運営以外のプレイヤーを非表示にします。")
                            itemStack.itemMeta = itemMeta

                            inv.setItem(8, itemStack)
                        }
                        else -> {
                            val worldConfig = Main.worldConfig
                            val difficultyConfig = worldConfig.getDifficultyConfig(p)

                            if (MapUtility.map != null) {
                                val mapStack = MapUtility.map!!.clone()
                                val mapMeta = mapStack.itemMeta!!
                                mapMeta.setDisplayName(ItemUtil.getItemName("地図"))
                                mapMeta.lore = listOf("${ChatColor.YELLOW}なんかすごいやつ (語彙力)")
                                mapStack.itemMeta = mapMeta

                                inv.setItem(40, mapStack)
                            }

                            val boneItem = difficultyConfig.getBone(type)
                            val boneStack = ItemStack(Material.BONE, boneItem.count)
                            val boneMeta = boneStack.itemMeta!!
                            boneMeta.setDisplayName(ItemUtil.getItemName("骨 (透明化)"))
                            boneMeta.lore = listOf(
                                    "${ChatColor.YELLOW}クリックで${DateTimeUtil.formatTimestamp(boneItem.duration).rawSeconds}秒間透明になります。",
                                    "",
                                    "${ChatColor.GREEN}${ChatColor.UNDERLINE}クールタイム${ChatColor.RESET}${ChatColor.GRAY}: ${ChatColor.YELLOW}${DateTimeUtil.formatTimestamp(boneItem.duration + 5).rawSeconds}${ChatColor.GRAY}秒"
                            )
                            boneStack.itemMeta = boneMeta

                            val featherItem = difficultyConfig.getFeather(type)
                            val featherStack = ItemStack(Material.FEATHER, featherItem.count)
                            val featherMeta = featherStack.itemMeta!!
                            featherMeta.setDisplayName(ItemUtil.getItemName("羽 (移動速度上昇)"))
                            featherMeta.lore = listOf(
                                    "${ChatColor.YELLOW}クリックで${DateTimeUtil.formatTimestamp(featherItem.duration).rawSeconds}秒間移動速度が上昇します。",
                                    "",
                                    "${ChatColor.GREEN}${ChatColor.UNDERLINE}クールタイム${ChatColor.RESET}${ChatColor.GRAY}: ${ChatColor.YELLOW}${DateTimeUtil.formatTimestamp(featherItem.duration + 5).rawSeconds}${ChatColor.GRAY}秒"
                            )
                            featherStack.itemMeta = featherMeta

                            val eggStack = ItemStack(Material.EGG, difficultyConfig.getEgg(type).count)
                            val eggMeta = eggStack.itemMeta!!
                            eggMeta.setDisplayName(ItemUtil.getItemName("卵 (盲目・移動速度低下)"))
                            eggMeta.lore = listOf("${ChatColor.YELLOW}ハンターに当てると盲目と移動速度低下を5秒間与えます。")
                            eggStack.itemMeta = eggMeta

                            val items = PlayerManager.loadConfig(p).inventoryConfig.items
                            inv.setItem(items[ItemType.BONE] ?: 0, boneStack)
                            inv.setItem(items[ItemType.FEATHER] ?: 1, featherStack)
                            inv.setItem(items[ItemType.EGG] ?: 2, eggStack)

                            MissionManager.setBook(p)
                        }
                    }
                } else {
                    if (p.isHunterGroup) {
                        setArmor(p)
                    } else {
                        MissionManager.setBook(p)
                    }
                }
            }
        }

        // 装備
        private fun setArmor(p: Player) {
            if (p.isHunterTeam) {
                p.inventory.clear()
                p.inventory.helmet = ItemStack(Material.DIAMOND_HELMET)
                p.inventory.chestplate = ItemStack(Material.DIAMOND_CHESTPLATE)
                p.inventory.leggings = ItemStack(Material.DIAMOND_LEGGINGS)
                p.inventory.boots = ItemStack(Material.DIAMOND_BOOTS)
            } else if (p.isTuhoTeam) {
                p.inventory.clear()

                val itemStack = ItemStack(Material.HONEYCOMB)
                val itemMeta = itemStack.itemMeta!!
                itemMeta.setCustomModelData(1002)
                itemStack.itemMeta = itemMeta

                p.inventory.helmet = itemStack
                p.inventory.chestplate = ItemStack(Material.GOLDEN_CHESTPLATE)
                p.inventory.leggings = ItemStack(Material.GOLDEN_LEGGINGS)
                p.inventory.boots = ItemStack(Material.GOLDEN_BOOTS)
            }
        }

        fun removeArmor(p: Player) {
            p.inventory.helmet = ItemStack(Material.AIR)
            p.inventory.chestplate = ItemStack(Material.AIR)
            p.inventory.leggings = ItemStack(Material.AIR)
            p.inventory.boots = ItemStack(Material.AIR)
        }

        // エフェクト
        fun setPotionEffect(p: Player, isRemove: Boolean = true) {
            if (isRemove)
                for (effect in p.activePotionEffects)
                    p.removePotionEffect(effect.type)

            p.addPotionEffect(PotionEffect(PotionEffectType.NIGHT_VISION, 200000, 1, false, false))

            if (p.isPlayerGroup) {
                if (GameManager.isGame()) {
                    p.addPotionEffect(PotionEffect(PotionEffectType.INVISIBILITY, 200, 1, false, false))
                    p.addPotionEffect(PotionEffect(PotionEffectType.SPEED, 200, 1, false, false))
                }
            } else if (p.isHunterTeam) {
                p.addPotionEffect(PotionEffect(PotionEffectType.SPEED, 200000, 0, false, false))
            }
        }

        fun hidePlayer(p: Player, time: Long = 20 * 10) {
            if (!p.isPlayerGroup) return

            VisibilityManager.add(p, VisibilityManager.VisibilityType.ITEM)
            Bukkit.getScheduler().runTaskLater(Main.pluginInstance, Runnable { VisibilityManager.remove(p, VisibilityManager.VisibilityType.ITEM) }, time)
        }


        // 権限管理
        fun isAdmin(p: Player): Boolean {
            return p.isOp && (isBroadCaster(p) || hasPermission(p) || p.isAdminTeam)
        }

        fun hasPermission(p: Player): Boolean {
            return PlayerManager.loadConfig(p).permission
        }

        fun isBroadCaster(p: Player): Boolean {
            return PlayerManager.loadConfig(p).broadCaster
        }

        fun addOp(p: Player) {
            if (p.isAdminTeam || hasPermission(p) || isBroadCaster(p))
                p.isOp = true
        }

        fun removeOp(p: Player) {
            if (!p.isAdminTeam || hasPermission(p) || isBroadCaster(p))
                p.isOp = false
        }

        // テレポート
        fun teleport(p: Player, location: Location) {
            Bukkit.getScheduler().runTask(Main.pluginInstance, Runnable { p.teleport(location) })
        }

        fun teleport(p: Player, collection: Collection<Location>) {
            if (collection.isEmpty()) return
            Bukkit.getScheduler().runTask(Main.pluginInstance, Runnable { p.teleport(collection.shuffled(ThreadLocalRandom.current())[0]) })
        }

        // 通知・サウンド
        fun sendInformationText(p: Player) {
            val team = Teams.getJoinedTeam(p)
            val difficulty = DifficultyManager.getDifficulty(p)

            val (color, name) = when (PlayerManager.loadConfig(p).bookForegroundColor) {
                PlayerConfig.BookForegroundColor.BLACK -> PlayerConfig.BookForegroundColor.BLACK to "黒"
                PlayerConfig.BookForegroundColor.WHITE -> PlayerConfig.BookForegroundColor.WHITE to "白"
            }

            p.sendMessage("""
                ${MainAPI.getPrefix(MainAPI.PrefixType.WARNING)}現在の設定
                ${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}チーム: ${team.color}${ChatColor.UNDERLINE}${ChatColor.stripColor(team.displayName)}
                ${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}難易度: ${difficulty.color}${ChatColor.UNDERLINE}${ChatColor.stripColor(difficulty.displayName)}
                ${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}レート: ${ChatColor.YELLOW}${ChatColor.UNDERLINE}${MoneyManager.getRate(p)}${ChatColor.RESET}${ChatColor.GRAY}円/秒
                ${MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY)}本の文字色: ${color.color}${ChatColor.UNDERLINE}$name
            """.trimIndent())
        }

        fun sendNotificationSound() {
            for (p in Bukkit.getOnlinePlayers())
                sendNotificationSound(p)
        }

        private fun sendNotificationSound(p: Player) {
            /*
            new BukkitRunnable() {
                int count = 0;

                public void run() {
                    count++;
                    if (count == 10 || count == 11 || count == 12 || count == 13 || count == 14 || count == 15) {
                    } else if (count == 25) {
                        cancel();
                    } else {
                        p.playSound(p.getLocation(), Sound.BLOCK_NOTE_BLOCK_PLING, 2, 2);
                    }
                }
            }.runTaskTimer(Main.pluginInstance, 3, 3);
            */
            object : BukkitRunnable() {
                var c = 26

                override fun run() {
                    if (c in 18..26 || c in 1..9)
                        p.playSound(p.location, Sound.BLOCK_NOTE_BLOCK_PLING, 1f, 2f)
                    if (c < 0)
                        cancel()
                    c--
                }
            }.runTaskTimerAsynchronously(Main.pluginInstance, 0, 2)
        }
    }

    enum class Objective(val objectName: String) {
        // サイドバー
        SIDEBAR("Toso_Sidebar"),

        // サイドバー 表示用チーム
        SIDEBAR_STATUS("T_SB_Status"),
        SIDEBAR_TEAM("T_SB_Team"),
        SIDEBAR_DIFFICULTY("T_SB_Difficulty"),
        SIDEBAR_REWARD("T_SB_Reward"),
        SIDEBAR_RATE("T_SB_Rate"),
        SIDEBAR_TEAM_PLAYER("T_SB_T_Player"),
        SIDEBAR_TEAM_HUNTER("T_SB_T_Hunter"),
        SIDEBAR_TEAM_JAIL("T_SB_T_Jail"),
        SIDEBAR_TEAM_SUCCESS("TSB_T_Success"),
        SIDEBAR_TEAM_TUHO("T_SB_T_Tuho"),

        // チーム
        TEAM_ADMIN("Toso_Admin"),
        TEAM_PLAYER("Toso_Player"),
        TEAM_HUNTER("Toso_Hunter"),
        TEAM_JAIL("Toso_Jail"),
        TEAM_SUCCESS("Toso_Success"),
        TEAM_TUHO("Toso_Tuho");

    }
}